File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/xpath/JPElement.java
package com.github.jmlparser.xpath;

import com.github.javaparser.metamodel.NodeMetaModel;
import com.github.javaparser.metamodel.PropertyMetaModel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.w3c.dom.*;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

/**
 * @author Alexander Weigl
 * @version 1 (30.11.22)
 */
class JPElement implements Element {
    public com.github.javaparser.ast.Node getAstNode() {
        return astNode;
    }

    final com.github.javaparser.ast.Node astNode;
    final NodeMetaModel meta;

    private final Element parent;
    int posInParent;

    @Nullable
    private List<JPElement> children = null;

    private List<Attr> attributes = null;


    JPElement(com.github.javaparser.ast.Node astNode, Element parent) {
        this.parent = parent;
        this.astNode = astNode;
        meta = astNode.getMetaModel();
    }

    @Nullable
    List<JPElement> lazyChildren() {
        if (children == null) {
            children = meta.getAllPropertyMetaModels()
                    .stream()
                    .filter(DocumentFactories::isNodeProperty)
                    .map(it -> it.getValue(this.astNode))
                    .flatMap(it -> {
                        if (it instanceof com.github.javaparser.ast.NodeList<?> seq) {
                            return seq.stream().map(
                                    n -> DocumentFactories.getElement(n, this));
                        } else if (it instanceof com.github.javaparser.ast.Node n) {
                            return Stream.of(DocumentFactories.getElement(n, this));
                        }
                        return null;
                    })
                    .map(JPElement.class::cast)
                    .toList();
            for (int i = 0; i < children.size(); i++) {
                children.get(i).posInParent = i;
            }
        }
        return children;
    }

    @NotNull
    List<Attr> lazyAttributes() {
        if (attributes == null) {
            var attr = meta.getAllPropertyMetaModels()
                    .stream()
                    .filter(it -> !DocumentFactories.isNodeProperty(it))
                    .map(it -> DocumentFactories.getAttribute(this, it))
                    .map(Attr.class::cast)
                    .toList();
            attributes = new ArrayList<>(attr);
            for (PseudoAttributeProvider provider : DocumentFactories.getAttributeProviders()) {
                var seq = provider.attributeForNode(astNode, this);
                if (seq != null) {
                    attributes.addAll(seq);
                }
            }
        }
        return attributes;
    }

    @Override
    public String getTagName() {
        return meta.getTypeName();
    }

    @NotNull
    @Override
    public String getAttribute(String name) {
        var prop = getProperty(name);
        return prop.getValue(astNode).toString();
    }

    PropertyMetaModel getProperty(String name) {
        return meta.getAllPropertyMetaModels().stream()
                .filter(it -> it.getName().equals(name))
                .findFirst().orElse(null);
    }

    @Override
    public void setAttribute(String name, String value) throws DOMException {
        throw new IllegalStateException();
    }

    @Override
    public void removeAttribute(String name) throws DOMException {
        throw new IllegalStateException();
    }

    @Override
    public Attr getAttributeNode(String name) {
        return lazyAttributes().stream()
                .filter(it -> it.getName().equals(name))
                .findFirst().orElse(null);
    }

    @Override
    public Attr setAttributeNode(Attr newAttr) throws DOMException {
        return null;
    }

    @Override
    public Attr removeAttributeNode(Attr oldAttr) throws DOMException {
        return null;
    }

    @NotNull
    @Override
    public NodeList getElementsByTagName(String name) {
        throw new IllegalStateException();
    }

    @NotNull
    @Override
    public String getAttributeNS(String namespaceURI, String localName) throws DOMException {
        throw new IllegalStateException();

    }

    @Override
    public void setAttributeNS(String namespaceURI, String qualifiedName, String value) throws DOMException {
        throw new IllegalStateException();

    }

    @Override
    public void removeAttributeNS(String namespaceURI, String localName) throws DOMException {
        throw new IllegalStateException();

    }

    @Override
    public Attr getAttributeNodeNS(String namespaceURI, String localName) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @Override
    public Attr setAttributeNodeNS(Attr newAttr) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @NotNull
    @Override
    public NodeList getElementsByTagNameNS(String namespaceURI, String localName) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @Override
    public boolean hasAttribute(String name) {
        return !meta.getAllPropertyMetaModels().isEmpty();
    }

    @Override
    public boolean hasAttributeNS(String namespaceURI, String localName) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @Override
    public TypeInfo getSchemaTypeInfo() {
        return null;
    }

    @Override
    public void setIdAttribute(String name, boolean isId) throws DOMException {

    }

    @Override
    public void setIdAttributeNS(String namespaceURI, String localName, boolean isId) throws DOMException {

    }

    @Override
    public void setIdAttributeNode(Attr idAttr, boolean isId) throws DOMException {

    }

    @NotNull
    @Override
    public String getNodeName() {
        return getTagName();
    }

    @Override
    public String getNodeValue() throws DOMException {
        return "";
    }

    @Override
    public void setNodeValue(String nodeValue) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @Override
    public short getNodeType() {
        return ELEMENT_NODE;
    }

    @Override
    public Node getParentNode() {
        return parent;
    }

    @NotNull
    @Override
    public NodeList getChildNodes() {
        return DocumentFactories.wrap(lazyChildren());
    }

    @Override
    public Node getFirstChild() {
        if (hasChildNodes())
            return lazyChildren().get(0);
        else
            return null;
    }

    @Override
    public Node getLastChild() {
        if (hasChildNodes())
            return lazyChildren().get(lazyChildren().size() - 1);
        else
            return null;
    }

    @Override
    public Node getPreviousSibling() {
        if (parent != null && posInParent != 0) {
            return parent.getChildNodes().item(posInParent - 1);
        }
        return null;
    }

    @Override
    public Node getNextSibling() {
        if (parent != null && posInParent + 1 < parent.getChildNodes().getLength()) {
            return parent.getChildNodes().item(posInParent + 1);
        }
        return null;
    }

    @Override
    public NamedNodeMap getAttributes() {
        return DocumentFactories.nodeMap(this);
    }

    @Override
    public Document getOwnerDocument() {
        return parent != null ? parent.getOwnerDocument() : null;
    }

    @Override
    public Node insertBefore(Node newChild, Node refChild) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @Override
    public Node replaceChild(Node newChild, Node oldChild) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @Override
    public Node removeChild(Node oldChild) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @Override
    public Node appendChild(Node newChild) throws DOMException {
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR, "");
    }

    @Override
    public boolean hasChildNodes() {
        return lazyChildren().size() != 0;
    }

    @Override
    public Node cloneNode(boolean deep) {
        return null;
    }

    @Override
    public void normalize() {

    }

    @Override
    public boolean isSupported(String feature, String version) {
        return false;
    }

    @Override
    public String getNamespaceURI() {
        return null;
    }

    @Override
    public String getPrefix() {
        return null;
    }

    @Override
    public void setPrefix(String prefix) throws DOMException {

    }

    @Override
    public String getLocalName() {
        return null;
    }

    @Override
    public boolean hasAttributes() {
        return false;
    }

    @Override
    public String getBaseURI() {
        return null;
    }

    @Override
    public short compareDocumentPosition(Node other) throws DOMException {
        return 0;
    }

    @Override
    public String getTextContent() throws DOMException {
        return null;
    }

    @Override
    public void setTextContent(String textContent) throws DOMException {

    }

    @Override
    public boolean isSameNode(Node other) {
        return false;
    }

    @Override
    public String lookupPrefix(String namespaceURI) {
        return null;
    }

    @Override
    public boolean isDefaultNamespace(String namespaceURI) {
        return false;
    }

    @Override
    public String lookupNamespaceURI(String prefix) {
        return null;
    }

    @Override
    public boolean isEqualNode(Node arg) {
        return false;
    }

    @Override
    public Object getFeature(String feature, String version) {
        return null;
    }

    @Override
    public Object setUserData(String key, Object data, UserDataHandler handler) {
        return null;
    }

    @Override
    public Object getUserData(String key) {
        return null;
    }
}

File: tools/xpath/src/main/kotlin/io/github/jmltoolkit/xpath/JPElement.kt
package io.github.jmltoolkit.xpath

import com.github.javaparser.ast.Node
import com.github.javaparser.ast.NodeList
import com.github.javaparser.metamodel.NodeMetaModel
import com.github.javaparser.metamodel.PropertyMetaModel
import org.w3c.dom.*

/**
 * @author Alexander Weigl
 * @version 1 (30.11.22)
 */
class JPElement(val astNode: Node, private val parent: Element) : Element {
    private val meta: NodeMetaModel = astNode.metaModel
    private var posInParent: Int = 0

    val children: List<JPElement> by lazy {
        meta.allPropertyMetaModels
            .filter(DocumentFactories::isNodeProperty)
            .map { it.getValue(this.astNode) }
            .flatMap {
                when (it) {
                    is NodeList<*> -> it.map { n -> DocumentFactories.getElement(n, this) }
                    is Node -> listOf(DocumentFactories.getElement(it, this))
                    else -> listOf()
                }
            }
            .map { it }
            .also {
                for (i in children.indices) {
                    children[i].posInParent = i
                }
            }
    }

    val attributes: List<Attr> by lazy {
        val attr = meta.allPropertyMetaModels
            .filter { !DocumentFactories.isNodeProperty(it) }
            .map { DocumentFactories.getAttribute(this, it) }
            .toMutableList()

        for (provider in DocumentFactories.attributeProviders) {
            val seq = provider.attributeForNode(astNode, this)
            attr.addAll(seq)
        }
        attr
    }

    override fun getTagName(): String = meta.typeName

    override fun getAttribute(name: String): String {
        val prop = getProperty(name)
        return prop.getValue(astNode).toString()
    }

    fun getProperty(name: String): PropertyMetaModel {
        return meta.allPropertyMetaModels.stream()
            .filter { it: PropertyMetaModel -> it.name == name }
            .findFirst().orElse(null)
    }

    @Throws(DOMException::class)
    override fun setAttribute(name: String, value: String) {
        throw IllegalStateException()
    }

    @Throws(DOMException::class)
    override fun removeAttribute(name: String) {
        throw IllegalStateException()
    }

    override fun getAttributeNode(name: String) = attributes.asSequence()
        .filter { it: Attr -> it.name == name }
        .firstOrNull()

    @Throws(DOMException::class)
    override fun setAttributeNode(newAttr: Attr): Attr? = null

    @Throws(DOMException::class)
    override fun removeAttributeNode(oldAttr: Attr): Attr? = null

    override fun getElementsByTagName(name: String): org.w3c.dom.NodeList {
        throw IllegalStateException()
    }

    @Throws(DOMException::class)
    override fun getAttributeNS(namespaceURI: String, localName: String): String {
        throw IllegalStateException()
    }

    @Throws(DOMException::class)
    override fun setAttributeNS(namespaceURI: String, qualifiedName: String, value: String) {
        throw IllegalStateException()
    }

    @Throws(DOMException::class)
    override fun removeAttributeNS(namespaceURI: String, localName: String) {
        throw IllegalStateException()
    }

    @Throws(DOMException::class)
    override fun getAttributeNodeNS(namespaceURI: String, localName: String): Attr {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    @Throws(DOMException::class)
    override fun setAttributeNodeNS(newAttr: Attr): Attr {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    @Throws(DOMException::class)
    override fun getElementsByTagNameNS(namespaceURI: String, localName: String): org.w3c.dom.NodeList {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    override fun hasAttribute(name: String): Boolean = !meta.allPropertyMetaModels.isEmpty()

    @Throws(DOMException::class)
    override fun hasAttributeNS(namespaceURI: String, localName: String): Boolean {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    override fun getSchemaTypeInfo(): TypeInfo? = null

    @Throws(DOMException::class)
    override fun setIdAttribute(name: String, isId: Boolean) {
    }

    @Throws(DOMException::class)
    override fun setIdAttributeNS(namespaceURI: String, localName: String, isId: Boolean) {
    }

    @Throws(DOMException::class)
    override fun setIdAttributeNode(idAttr: Attr, isId: Boolean) {
    }

    override fun getNodeName(): String = tagName

    @Throws(DOMException::class)
    override fun getNodeValue(): String = ""

    @Throws(DOMException::class)
    override fun setNodeValue(nodeValue: String) {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    override fun getNodeType(): Short = org.w3c.dom.Node.ELEMENT_NODE

    override fun getParentNode(): org.w3c.dom.Node = parent

    override fun getChildNodes(): org.w3c.dom.NodeList = DocumentFactories.wrap(children)

    override fun getFirstChild() = children.firstOrNull()
    override fun getLastChild() = children.lastOrNull()

    override fun getPreviousSibling(): org.w3c.dom.Node? {
        if (posInParent != 0) {
            return parent.childNodes.item(posInParent - 1)
        }
        return null
    }

    override fun getNextSibling(): org.w3c.dom.Node? {
        if (posInParent + 1 < parent.childNodes.length) {
            return parent.childNodes.item(posInParent + 1)
        }
        return null
    }

    override fun getAttributes(): NamedNodeMap = DocumentFactories.nodeMap(this)

    override fun getOwnerDocument(): Document? = parent.ownerDocument

    @Throws(DOMException::class)
    override fun insertBefore(newChild: org.w3c.dom.Node, refChild: org.w3c.dom.Node): org.w3c.dom.Node {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    @Throws(DOMException::class)
    override fun replaceChild(newChild: org.w3c.dom.Node, oldChild: org.w3c.dom.Node): org.w3c.dom.Node {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    @Throws(DOMException::class)
    override fun removeChild(oldChild: org.w3c.dom.Node): org.w3c.dom.Node {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    @Throws(DOMException::class)
    override fun appendChild(newChild: org.w3c.dom.Node): org.w3c.dom.Node {
        throw DOMException(DOMException.NOT_SUPPORTED_ERR, "")
    }

    override fun hasChildNodes(): Boolean = children.isNotEmpty()

    override fun cloneNode(deep: Boolean): org.w3c.dom.Node? = null

    override fun normalize() {
    }

    override fun isSupported(feature: String, version: String): Boolean = false

    override fun getNamespaceURI(): String? = null

    override fun getPrefix(): String? = null

    @Throws(DOMException::class)
    override fun setPrefix(prefix: String) {
    }

    override fun getLocalName(): String? = null

    override fun hasAttributes(): Boolean = false

    override fun getBaseURI(): String? = null

    @Throws(DOMException::class)
    override fun compareDocumentPosition(other: org.w3c.dom.Node): Short = 0

    @Throws(DOMException::class)
    override fun getTextContent(): String? = null

    @Throws(DOMException::class)
    override fun setTextContent(textContent: String) {
    }

    override fun isSameNode(other: org.w3c.dom.Node): Boolean = false

    override fun lookupPrefix(namespaceURI: String): String? = null

    override fun isDefaultNamespace(namespaceURI: String): Boolean = false

    override fun lookupNamespaceURI(prefix: String): String? = null

    override fun isEqualNode(arg: org.w3c.dom.Node): Boolean = false

    override fun getFeature(feature: String, version: String): Any? = null

    override fun setUserData(key: String, data: Any, handler: UserDataHandler): Any? = null

    override fun getUserData(key: String): Any? = null
}
